providers/implementations/encode_decode/encode_key2text.c

```diff
@@ -0,0 +1,884 @@
+                    goto err;
+            }
+            break;
+        case RSA_FLAG_TYPE_RSASSAPSS:
+            if (rsa_pss_params_30_is_unrestricted(pss_params)) {
+                if (BIO_printf(out, "No PSS parameter restrictions\n") <= 0)
+                    goto err;
+            } else {
+                int hashalg_nid = rsa_pss_params_30_hashalg(pss_params);
+                int maskgenalg_nid = rsa_pss_params_30_maskgenalg(pss_params);
+                int maskgenhashalg_nid =
+                    rsa_pss_params_30_maskgenhashalg(pss_params);
+                int saltlen = rsa_pss_params_30_saltlen(pss_params);
+                int trailerfield = rsa_pss_params_30_trailerfield(pss_params);
+
+                if (BIO_printf(out, "PSS parameter restrictions:\n") <= 0)
+                    goto err;
+                if (BIO_printf(out, "  Hash Algorithm: %s%s\n",
+                               rsa_oaeppss_nid2name(hashalg_nid),
+                               (hashalg_nid == NID_sha1
+                                ? " (default)" : "")) <= 0)
+                    goto err;
+                if (BIO_printf(out, "  Mask Algorithm: %s with %s%s\n",
+                               rsa_mgf_nid2name(maskgenalg_nid),
+                               rsa_oaeppss_nid2name(maskgenhashalg_nid),
+                               (maskgenalg_nid == NID_mgf1
+                                && maskgenhashalg_nid == NID_sha1
+                                ? " (default)" : "")) <= 0)
+                    goto err;
+                if (BIO_printf(out, "  Minimum Salt Length: %d%s\n",
+                               saltlen,
+                               (saltlen == 20 ? " (default)" : "")) <= 0)
+                    goto err;
+                /*
+                 * TODO(3.0) Should we show the ASN.1 trailerField value, or
+                 * the actual trailerfield byte (i.e. 0xBC for 1)?
+                 * crypto/rsa/rsa_ameth.c isn't very clear on that, as it
+                 * does display 0xBC when the default applies, but the ASN.1
+                 * trailerField value otherwise...
+                 */
+                if (BIO_printf(out, "  Trailer Field: 0x%x%s\n",
+                               trailerfield,
+                               (trailerfield == 1 ? " (default)" : "")) <= 0)
+                    goto err;
+            }
+            break;
+        }
+    }
+
+    ret = 1;
+ err:
+    sk_BIGNUM_const_free(factors);
+    sk_BIGNUM_const_free(exps);
+    sk_BIGNUM_const_free(coeffs);
+    return ret;
+}
+
+/* ---------------------------------------------------------------------- */
+
+static void *key2text_newctx(void *provctx)
+{
+    return provctx;
+}
+
+static void key2text_freectx(void *vctx)
+{
+}
+
+static int key2text_encode(void *vctx, const void *key, int selection,
+                           OSSL_CORE_BIO *cout,
+                           int (*key2text)(BIO *out, const void *key,
```

Comment:
can this be typedefed and simplfied?

Reply:
Typedefed for one time use?  Why?
I'm not sure what you mean with "simplify"